### PR TITLE
History Retrieval Schema

### PR DESCRIPTION
**Pull request by ingalls**
### Context

We currently support full feature history retrieval that is based on the stored FeatureCollection within a given delta.

Although this allows us to return a full geometry for any given feature in time, it is very slow due to having to iterate through all the features in a given delta to find the requested feature.

This PR scopes breaking out the features in a given delta to be contained in their own table.

The new feature history query would then be able to simply select the feature history out of a single table with no additional parsing

### Feature History

```SQL
SELECT
  *
FROM
  geo_history
WHERE
  id = 123
ORDER BY
  VERSION DESC
```

### Delta Retrieval

```SQL
SELECT
  deltas.id,
  deltas.uid,
  users.username,
  geo_history.id,
  deltas.affected,
  deltas.props,
  deltas.created,
  deltas.props
 FROM
  deltas
  INNER JOIN users
    ON deltas.uid = users.id,
  geo_history
WHERE
  deltas.id = geo_history.delta
```

cc/ @ingalls

### CODE REVIEW

**Comment by lizziegooding** [comment]
Posting the query plans/execution time for the current feature history query vs the new proposed feature history query:

### Current schema
```sql
QUERY PLAN

--------------------------------------------------------------------------------------------------------------------------------------------
-------------------------
Aggregate  (cost=3942.07..3942.08 rows=1 width=32) (actual time=99.199..99.199 rows=1 loops=1)
->  Subquery Scan on t  (cost=2608.76..3940.75 rows=264 width=49) (actual time=25.665..99.173 rows=3 loops=1)
Filter: (((t.feat ->> 'id'::text))::bigint = 123)
Rows Removed by Filter: 11756
->  ProjectSet  (cost=2608.76..2886.75 rows=52700 width=57) (actual time=24.044..68.504 rows=11759 loops=1)
->  Result  (cost=2608.76..2615.35 rows=527 width=45) (actual time=0.462..0.467 rows=3 loops=1)
->  Sort  (cost=2608.76..2610.08 rows=527 width=45) (actual time=0.461..0.463 rows=3 loops=1)
Sort Key: deltas.id DESC
Sort Method: quicksort  Memory: 25kB
->  Hash Join  (cost=1310.40..2584.93 rows=527 width=45) (actual time=0.445..0.452 rows=3 loops=1)
Hash Cond: (deltas.uid = users.id)
->  Bitmap Heap Scan on deltas  (cost=1308.09..2575.37 rows=527 width=36) (actual time=0.413..0.416 rows=3
loops=1)
Recheck Cond: (affected @> '{123}'::bigint[])
Heap Blocks: exact=3
->  Bitmap Index Scan on deltas_affected_idx  (cost=0.00..1307.95 rows=527 width=0) (actual time=0.40
8..0.408 rows=3 loops=1)
Index Cond: (affected @> '{123}'::bigint[])
->  Hash  (cost=2.14..2.14 rows=14 width=17) (actual time=0.024..0.024 rows=14 loops=1)
Buckets: 1024  Batches: 1  Memory Usage: 9kB
->  Seq Scan on users  (cost=0.00..2.14 rows=14 width=17) (actual time=0.008..0.015 rows=14 loops=1)
Planning Time: 0.218 ms
Execution Time: 99.391 ms
```

### New schema 
> The new feature history query would then be able to simply select the feature history out of a single table with no additional parsing

This is not exactly true-- it still needs to pull data from the `deltas` and `users` table and convert everything to JSON. Here's the query i'm using now to replicate our existing functionality:
```sql
SELECT json_agg (
    JSON_Build_Object(
        'id', deltas.id,
        'uid', deltas.uid,
        'feat', JSON_Build_Object(
            'id', geo_history.id,
            'action', geo_history.action,
            'key', geo_history.key,
            'type', 'Feature',
            'version', geo_history.version,
            'geometry', ST_AsGeoJSON(geom)::JSON,
            'properties', geo_history.props
        ),
        'username', users.username
    )
)
FROM
    geo_history,
    deltas,
    users
WHERE 
    geo_history.id = 366672 AND
    geo_history.delta = deltas.id AND
    deltas.uid = users.id
```
I'd welcome suggestions on how to make this more performant.

```sql
                                                                 QUERY PLAN
--------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=188344.67..188344.68 rows=1 width=32) (actual time=3.987..3.987 rows=1 loops=1)
   ->  Hash Join  (cost=1474.94..113572.67 rows=29760 width=192) (actual time=3.918..3.925 rows=2 loops=1)
         Hash Cond: (geo_history.delta = deltas.id)
         ->  Bitmap Heap Scan on geo_history  (cost=1411.48..112342.15 rows=46312 width=152) (actual time=0.023..0.026 rows=2 loops=1)
               Recheck Cond: (id = 366672)
               Heap Blocks: exact=2
               ->  Bitmap Index Scan on geo_history_pkey  (cost=0.00..1399.90 rows=46312 width=0) (actual time=0.018..0.018 rows=2 loops=1)
                     Index Cond: (id = 366672)
         ->  Hash  (cost=61.85..61.85 rows=129 width=48) (actual time=3.887..3.887 rows=1500 loops=1)
               Buckets: 2048 (originally 1024)  Batches: 1 (originally 1)  Memory Usage: 99kB
               ->  Hash Join  (cost=1.32..61.85 rows=129 width=48) (actual time=0.042..2.890 rows=1500 loops=1)
                     Hash Cond: (deltas.uid = users.id)
                     ->  Seq Scan on deltas  (cost=0.00..52.36 rows=1836 width=16) (actual time=0.009..0.955 rows=1500 loops=1)
                     ->  Hash  (cost=1.14..1.14 rows=14 width=40) (actual time=0.026..0.026 rows=14 loops=1)
                           Buckets: 1024  Batches: 1  Memory Usage: 9kB
                           ->  Seq Scan on users  (cost=0.00..1.14 rows=14 width=40) (actual time=0.005..0.014 rows=14 loops=1)
 Planning time: 0.180 ms
 Execution time: 4.046 ms
```

It does still improve query time by a factor of 25, `99.391 ms` vs `4.046 ms`

File: src/schema.sql [hunk 1]
```diff
@@ -5,28 +5,44 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto;
...
 DROP TABLE IF EXISTS webhooks;
+DROP TABLE IF EXISTS meta;
+DROP TABLE IF EXISTS users_tokens;
+DROP TABLE IF EXISTS bounds;
+DROP TABLE IF EXISTS tiles;
+DROP TABLE IF EXISTS geo;
+DROP TABLE IF EXISTS geo_history;
+DROP TABLE IF EXISTS styles;
+DROP TABLE IF EXISTS deltas;
+DROP TABLE IF EXISTS users;
```
Comment (miccolis) [inline src/schema.sql]: This  re-org makes the PR hard to follow. Can we unroll this, or at least get the changes into a single commit that can be viewed on it's own?

File: src/schema.sql [hunk 2]
```diff
@@ -5,28 +5,44 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto;
...
+
+CREATE TABLE users (
+    id          BIGSERIAL PRIMARY KEY,
+    access      TEXT,
+    username    TEXT UNIQUE NOT NULL,
+    password    TEXT NOT NULL,
+    email       TEXT UNIQUE NOT NULL,
+    meta        JSONB
+);
```
Comment (miccolis) [inline src/schema.sql]: Same here, why is this table being moved? Can it stay in place?

File: src/schema.sql [hunk 3]
```diff
@@ -5,28 +5,44 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto;
...
+
 CREATE TABLE webhooks (
     id          BIGSERIAL PRIMARY KEY,
     name        TEXT NOT NULL,
-    actions     TEXT[],
+    actions     TEXT[] NOT NULL,
```
Comment (miccolis) [inline src/schema.sql]: Changes seems unrelated to the 'history retrieval schema' I think we should move it to another PR.

File: src/schema.sql [hunk 4]
```diff
@@ -36,30 +52,28 @@ CREATE TABLE bounds (
...
-    password    TEXT NOT NULL,
-    email       TEXT UNIQUE NOT NULL,
-    meta        JSONB
-);
-
-DROP TABLE IF EXISTS users_tokens;
 CREATE TABLE users_tokens (
     name        TEXT,
-    uid         BIGINT,
+    uid         BIGINT REFERENCES users(id),
```
Comment (miccolis) [inline src/schema.sql]: Changes seems unrelated to the 'history retrieval schema' I think we should move it to another PR.

File: src/schema.sql [hunk 5]
```diff
@@ -69,28 +83,27 @@ CREATE TABLE geo (
...
+CREATE INDEX geo_history_gist ON geo_history USING GIST(geom);
+CREATE INDEX geo_history_idx ON geo_history(id);
+CREATE INDEX geo_history_deltax ON geo_history(delta);
+
 CREATE TABLE styles (
     id          BIGSERIAL PRIMARY KEY,
     name        TEXT NOT NULL,
     style       JSONB NOT NULL,
-    uid         BIGINT NOT NULL,
+    uid         BIGINT NOT NULL REFERENCES users(id),
```
Comment (miccolis) [inline src/schema.sql]: Changes seems unrelated to the 'history retrieval schema' I think we should move it to another PR.

File: src/schema.sql [hunk 6]
```diff
@@ -36,30 +52,28 @@ CREATE TABLE bounds (
...
-DROP TABLE IF EXISTS geo;
-DROP INDEX IF EXISTS geo_gist;
-DROP INDEX IF EXISTS geo_idx;
+CREATE TABLE deltas (
+    id          BIGSERIAL PRIMARY KEY,
+    created     TIMESTAMP,
+    features    JSONB,
+    affected    BIGINT[],
+    props       JSONB,
+    uid         BIGINT REFERENCES users(id),
```
Comment (miccolis) [inline src/schema.sql]: It's unclear to me that we really need a foreign key constraint here. We don't have them elsewhere, whats the thinking here?

File: src/schema.sql [hunk 7]
```diff
@@ -69,28 +83,27 @@ CREATE TABLE geo (
 CREATE INDEX geo_gist ON geo USING GIST(geom);
 CREATE INDEX geo_idx ON geo(id);
 
-DROP TABLE IF EXISTS styles;
+CREATE TABLE geo_history (
+    id          BIGINT NOT NULL,
+    delta       BIGINT NOT NULL REFERENCES deltas(id),
```
Comment (miccolis) [inline src/schema.sql]: I understand the application of the reference here, but I'd like to know more about what we're trying to protect against by adding it. The constraint will add time to inserts, etc...

File: src/schema.sql [hunk 8]
```diff
@@ -36,30 +52,28 @@ CREATE TABLE bounds (
...
+    props       JSONB,
+    uid         BIGINT REFERENCES users(id),
+    finalized   BOOLEAN DEFAULT FALSE
+);
+CREATE INDEX deltas_idx ON deltas(id);
+CREATE INDEX deltas_affected_idx on deltas USING GIN (affected);
+
 CREATE TABLE geo (
-    id          BIGSERIAL UNIQUE,
+    id          BIGSERIAL PRIMARY KEY,
```
Comment (miccolis) [inline src/schema.sql]: `UNIQUE` and `PRIMARY KEY` are pretty similar (identical?), what's the reasoning here?

**Review by miccolis — DISMISSED** [review]
I really like the direction here, but I do have some requests. It would be great to keep schema changes focused on history retrieval only. Left a bunch of comments to that end below.

File: src/schema.sql (hunk 8 above)
Comment (lizziegooding) [reply src/schema.sql]: The difference is that `UNIQUE` allows one or more `NULL` values whereas `PRIMARY KEY` enforces both a `UNIQUE` and a `NOT NULL` constraint. I originally added this as it was required by Diesel, but per chat with @ingalls, is probably something we should add anyway (though we don't have to do it in this PR)

File: src/schema.sql (hunk 1 above)
Comment (ingalls) [reply src/schema.sql]: @miccolis Can't unroll this now as due to the addition of proper references, this needs to be at the top and in the correct order or schema application will fail. I'll defer to @lizziegooding for rolling into a single commit

File: src/schema.sql (hunk 2 above)
Comment (ingalls) [reply src/schema.sql]: No, it needs to be moved to be created as other referenced tables now depend on it

File: src/schema.sql (hunk 6 above)
Comment (ingalls) [reply src/schema.sql]: Not having them was an oversight on my part, we should definitely have references where possible to maintain database integrity.

File: src/delta/mod.rs [hunk 9]
```diff
@@ -38,32 +38,40 @@ impl Delta {
 }
 
 ///Get the history of a particular feature
+/// @TODO move to feature modules?
```
Comment (ingalls) [inline src/delta/mod.rs]: 100% agree, let's do this, or create a new history mod

File: src/feature/mod.rs [hunk 10]
```diff
@@ -283,16 +283,44 @@ pub fn create(trans: &postgres::transaction::Transaction, schema: &Option<valico
...
-                old: id,
-                new: Some(res.get(0).get(0)),
-                version: Some(1)
-            }),
+            Ok(res) => {
+                let new: i64 = res.get(0).get(0);
+                match trans.query("
+                    INSERT INTO geo_history (version, action, geom, props, delta, key, id)
+                        VALUES (
+                            1,
```
Comment (ingalls) [inline src/feature/mod.rs]: can we just remove the `key` unique value from the geo_history column? As this doesn't/shouldn't be unique on the geo_history column (I realize I probably had it set when I created it based on the geo table)

Comment (lizziegooding) [reply src/feature/mod.rs]: Addressed in https://github.com/mapbox/Hecate/pull/206/commits/e1baff273e876d047cdc788d7aa4dcf4c950463f

File: src/schema.sql (hunk 1 above)
Comment (lizziegooding) [reply src/schema.sql]: Removed in https://github.com/mapbox/Hecate/pull/206/commits/7493e0efafaeec0af7b70cbab56df3f1b16525d5 -- while these are changes we should make, i'm removing them from the current PR to reduce scope for what will already be a large change.

File: src/schema.sql (hunk 2 above)
Comment (lizziegooding) [reply src/schema.sql]: Removed in https://github.com/mapbox/Hecate/pull/206/commits/7493e0efafaeec0af7b70cbab56df3f1b16525d5 -- while these are changes we should make, i'm removing them from the current PR to reduce scope for what will already be a large change.

File: src/schema.sql (hunk 3 above)
Comment (lizziegooding) [reply src/schema.sql]: Removed in https://github.com/mapbox/Hecate/pull/206/commits/7493e0efafaeec0af7b70cbab56df3f1b16525d5 -- while these are changes we should make, i'm removing them from the current PR to reduce scope for what will already be a large change.

File: src/schema.sql (hunk 4 above)
Comment (lizziegooding) [reply src/schema.sql]: Removed in https://github.com/mapbox/Hecate/pull/206/commits/7493e0efafaeec0af7b70cbab56df3f1b16525d5 -- while these are changes we should make, i'm removing them from the current PR to reduce scope for what will already be a large change.

File: src/schema.sql (hunk 5 above)
Comment (lizziegooding) [reply src/schema.sql]: Removed in https://github.com/mapbox/Hecate/pull/206/commits/7493e0efafaeec0af7b70cbab56df3f1b16525d5 -- while these are changes we should make, i'm removing them from the current PR to reduce scope for what will already be a large change.

File: src/schema.sql (hunk 6 above)
Comment (lizziegooding) [reply src/schema.sql]: Removed in https://github.com/mapbox/Hecate/pull/206/commits/7493e0efafaeec0af7b70cbab56df3f1b16525d5 -- while these are changes we should make, i'm removing them from the current PR to reduce scope for what will already be a large change.

File: src/schema.sql (hunk 7 above)
Comment (lizziegooding) [reply src/schema.sql]: Removed in https://github.com/mapbox/Hecate/pull/206/commits/7493e0efafaeec0af7b70cbab56df3f1b16525d5 -- while these are changes we should make, i'm removing them from the current PR to reduce scope for what will already be a large change.

File: src/delta/mod.rs [hunk 11]
```diff
@@ -37,39 +37,6 @@ impl Delta {
     }
 }
 
-///Get the history of a particular feature
-pub fn history(conn: &impl postgres::GenericConnection, feat_id: &i64) -> Result<serde_json::Value, HecateError> {
```
Comment (lizziegooding) [inline src/delta/mod.rs:41]: Moved to src/feature/mod.rs

File: src/feature/mod.rs [hunk 12]
```diff
@@ -525,46 +635,41 @@ pub fn restore(trans: &postgres::transaction::Transaction, schema: &Option<valic
...
-        Ok(history) => {
+        Ok(res) => {
+            // Agg function returns row of NULLs if inner query doesn't return results
+            let affected: Option<Vec<i64>> = res.get(0).get(0);
+            let affected = match affected {
+                None => return Err(import_error(&feat, "Feature Not Found", None)),
+                Some(deltas) => deltas
+            };
 
-            if history.len() != 1 {
```
Comment (lizziegooding) [inline src/feature/mod.rs:549]: In the new aggregate function, this was not catching when no results were returned from the query

File: src/feature/mod.rs [hunk 13]
```diff
@@ -525,46 +635,41 @@ pub fn restore(trans: &postgres::transaction::Transaction, schema: &Option<valic
...
-                Some(prev_version) => {
-                    if prev_version != version {
-                        return Err(import_error(&feat, "Restore Version Mismatch", None));
-                    }
-                }
-            };
-
-            let affected: Vec<i64> = history.get(0).get(0);
+            let prev_action: String = res.get(0).get(2);
+            if prev_action != String::from("delete") {
```
Comment (lizziegooding) [inline src/feature/mod.rs:670]: This now checks that the last action was a `delete` rather than just checking that the feature's most recent modification was a `create`

File: src/feature/mod.rs [hunk 14]
```diff
@@ -576,22 +681,47 @@ pub fn restore(trans: &postgres::transaction::Transaction, schema: &Option<valic
...
+                            }
+                        }
+                    }
+                },
                 Err(err) => {
                     match err.as_db() {
                         Some(e) => {
-                            if e.message == "duplicate key value violates unique constraint \"geo_id_key\"" {
+                            if e.message == "duplicate key value violates unique constraint \"geo_id_key\"" ||
+                            e.message == "duplicate key value violates unique constraint \"geo_pkey\"" {
```
Comment (lizziegooding) [inline src/feature/mod.rs]: Added to account for new primary key constraint

File: src/feature/mod.rs [hunk 15]
```diff
@@ -694,3 +824,95 @@ pub fn get_bbox(conn: &impl postgres::GenericConnection, bbox: Vec<f64>) -> Resu
...
+            }
+            let history: serde_json::Value = res.get(0).get(0);
+
+            Ok(history)
+        },
+        Err(err) => Err(HecateError::from_db(err))
+    }
+}
+
+pub fn get_point_history_stream(conn: r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>, point: &String) -> Result<PGStream, HecateError> {
```
Comment (lizziegooding) [inline src/feature/mod.rs:868]: Based on `get_point_stream`

File: src/feature/mod.rs [hunk 16]
```diff
@@ -694,3 +824,95 @@ pub fn get_bbox(conn: &impl postgres::GenericConnection, bbox: Vec<f64>) -> Resu
...
+                FROM geo_history
+                WHERE
+                    ST_DWithin(ST_SetSRID(ST_MakePoint($1, $2), 4326), geom, 0.00005)
+                ORDER BY
+                    ST_Distance(ST_SetSRID(ST_MakePoint($1, $2), 4326), geom) DESC
+            ) f;
+    "#), &[&lng, &lat])?)
+}
+
+pub fn get_bbox_history_stream(conn: r2d2::PooledConnection<r2d2_postgres::PostgresConnectionManager>, bbox: &Vec<f64>) -> Result<PGStream, HecateError> {
```
Comment (lizziegooding) [inline src/feature/mod.rs:894]: Based on `get_bbox_stream`

File: tests/history.rs [hunk 17]
```diff
@@ -70,7 +70,7 @@ mod test {
 
         {
             let mut resp = reqwest::get("http://localhost:8000/api/data/feature/1/history").unwrap();
-            assert_eq!(resp.text().unwrap(), r#"[{"feat":{"action":"create","geometry":{"coordinates":[0.0,0.0],"type":"Point"},"id":1,"message":"Creating a Point","properties":{"number":"123"},"type":"Feature"},"id":1,"uid":1,"username":"ingalls"}]"#);
+            assert_eq!(resp.text().unwrap(), r#"[{"feat":{"action":"create","geometry":{"coordinates":[0,0],"type":"Point"},"id":1,"key":null,"properties":{"number":"123"},"type":"Feature","version":1},"id":1,"uid":1,"username":"ingalls"}]"#);
```
Comment (lizziegooding) [inline tests/history.rs]: Some important differences between the previous feature history behavior and the current API:
- coordinates are stored in a `GEOMETRY` type rather than as pure JSON and as a result are simplified (e.g. `0.0` -> `0`)
- `key` values are always returned, even if they're not included in the original request
- the `version` returned reflects the actual feature version, not the version requested. For example, when a feature is created, it has a version of 1
- the `message` property which was only passed as a property on the feature when a single feature was created/deleted/modified/restored, is recorded in the `deltas` table but never returned from the history endpoint. Previously, it was included in the response when a single feature was created/deleted/modified/restored but not passed if multiple features were changed

File: src/schema.sql (hunk 8 above)
Comment (miccolis) [reply src/schema.sql]: I'd really prefer to keep a PR like this clean of unrelated changes. If this adds any time to the migration at all, I'm very against it. I totally get the desire to make changes like this, but I'd vote against schema maintenance in here.

Comment (lizziegooding) [reply src/schema.sql]: Amended previous commit to remove this

File: src/feature/mod.rs (hunk 14 above)
Comment (lizziegooding) [reply src/feature/mod.rs]: Removed, `geo` no longer has primary key constraint

**Comment by ingalls** [comment]
:taco:

**Review by lizziegooding — APPROVED** [review]
🌮 from @ingalls and @miccolis
